src/libmongoc/doc/structured_log.rst

```diff
@@ -0,0 +1,127 @@
+* A default handler is installed, which logs text representations of each message to a file given by ``MONGODB_LOG_PATH``, which may be a full path or one of the special values ``stdout`` or ``stderr``. If no valid path is given, logs are written to stderr.
+
+Normally environment variables provide defaults that can be overridden programmatically.
+To request the opposite behavior, where your programmatic defaults can be overridden by the environment, see :symbol:`mongoc_structured_log_set_max_levels_from_env`.
+
+Levels and Components
+---------------------
+
+Log levels and components are defined as :symbol:`mongoc_structured_log_level_t` and :symbol:`mongoc_structured_log_component_t` enumerations. Utilities are provided to convert between these values and their standard string representations. The string values are case-insensitive.
+
+.. code-block:: c
+
+  typedef enum {
+    MONGOC_STRUCTURED_LOG_LEVEL_EMERGENCY = 0,
+    MONGOC_STRUCTURED_LOG_LEVEL_ALERT = 1,
+    MONGOC_STRUCTURED_LOG_LEVEL_CRITICAL = 2,
+    MONGOC_STRUCTURED_LOG_LEVEL_ERROR = 3,
+    MONGOC_STRUCTURED_LOG_LEVEL_WARNING = 4,
+    MONGOC_STRUCTURED_LOG_LEVEL_NOTICE = 5,
+    MONGOC_STRUCTURED_LOG_LEVEL_INFO = 6,
+    MONGOC_STRUCTURED_LOG_LEVEL_DEBUG = 7,
+    MONGOC_STRUCTURED_LOG_LEVEL_TRACE = 8,
+  } mongoc_structured_log_level_t;
+
+  typedef enum {
+    MONGOC_STRUCTURED_LOG_COMPONENT_COMMAND = 0,
+    MONGOC_STRUCTURED_LOG_COMPONENT_TOPOLOGY = 1,
+    MONGOC_STRUCTURED_LOG_COMPONENT_SERVER_SELECTION = 2,
+    MONGOC_STRUCTURED_LOG_COMPONENT_CONNECTION = 3,
+  } mongoc_structured_log_component_t;
+
+.. toctree::
+  :titlesonly:
+  :maxdepth: 1
+
+  mongoc_structured_log_level_t
+  mongoc_structured_log_component_t
+
+.. seealso::
+
+  mongoc_structured_log_get_level_name
+  mongoc_structured_log_get_named_level
+  mongoc_structured_log_get_component_name
+  mongoc_structured_log_get_named_component
+
+
+Log Filtering
+-------------
+
+Structured log messages may be filtered in two ways:
+
+* A maximum log level can be set per-component.
+* A log handler function can ignore messages based on any criteria.
+
+The max level settings are configured by the environment variables above, and they may be altered or queried by applications.
+To reduce overhead for unwanted logging, messages should be ignored as early as possible.
+If it's possible to filter messages based on only their component and level, this should be done by handlers before requesting BSON serialization by calling :symbol:`mongoc_structured_log_entry_message_as_bson`.
```

Comment:
Yes, you can filter by log level either before the callback or inside. My thinking is that the ahead-of-time filtering is a performance optimization for the common case of setting a max level per component, but that the handler would allow any other type of filtering not covered by that.

For now the message 'envelope' (items that the handler can access without building a bson_t) only includes level and component, but at a small cost in design flexibility we could also choose to include the message string; that would make it possible to make a trace facility that saves the level, component, and message string from every log message but only the full bson data for specific messages.